lib/flatpak-installation.c

```diff
@@ -1121,6 +1121,31 @@ get_write_progress (guint outstanding_writes)
   return (guint) (3 / (gdouble) outstanding_writes);
 }
 
+static gboolean
+app_contains_extra_data (const char *app_id)
+{
+  /* HACK: This is a temporary solution until the detached metadata work is ready */
+  const char *extra_data_apps[] =
```

Comment:
Flatpak uses OSTree metadata to store the extra-data. Thus, it is only able to detect extra-data **after** the entire OSTree repo is pulled. Since we can't rely on that **while** the pull is happening, Flatpak can't detect it it has extra-data, not until it's too late.

Because we can't detect that while pulling, the pull progress goes up to 100% - then, when it finally detects extra-data, the progress already can't be rolled back.

If Flatpak used OSTree's detached metadata this wouldn't happen. I'm working on it as I write this comment, but still didn't managed to get something working. For 3.1.4, @rshuler proposed using this approach temporarily until the detached metadata solution is ready (for 3.2).